4750 message content

latin-panda opened the pull request:
# Description

This PR migrates:
- message content
- sender component
- send message modal
- delete modal
- count message feature

medic/cht-core#4750

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

latin-panda commented:
Hi @dianabarsan , can you please take a look? Thank you!!

dianabarsan reviewed (changes requested):
Nice work! 

I added a few comments inline, and besides those I'm seeing some inconsistencies compared with `master`:

1. I have a rather long LHS conversations list. When I select one and submit a message, I'm seeing the loader spinner when the LHS is updated, while on master, I don't see the spinner.
2. When selecting a really long conversation, scrolling to the top doesn't load more messages, I'm also not seeing "No more messages" when all messages are loaded. 
3. I'm not seeing the contact phone number when loading a conversation.
4. Items on the LHS get a weird border when selected. 
![image](https://user-images.githubusercontent.com/35681649/95725754-ba8bbf80-0c80-11eb-941d-057784db702d.png)

5. I'm seeing inconsistent results when searching in the select2search. 
![image](https://user-images.githubusercontent.com/35681649/95725889-e313b980-0c80-11eb-9565-95ac51b4c4fa.png)

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+
+  listTrackBy(index, message) {
+    const identifier = message.doc ? message.doc.id + message.doc._rev : message.id;
+    return message.key + identifier;
```
Yes, the "key" is the conversation, the "id" is unique per message I believe, so shouldn't be a problem in duplicates if the doc isn't coming. But I do see a bug, that's fixed now and I removed the key to keep it short.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+      recipient.doc = { contact: { phone: this.selectedConversation.id } };
+    }
+
+    this.sendMessageService.send(recipient, this.send.message)
```
Fixed.

latin-panda replied:
Fixed

latin-panda commented:
> 4. Items on the LHS get a weird border when selected.

This is weird, I didn't see it on Friday when I tested the router-link, I have added the `outline: none` and it's fixed.

latin-panda commented on webapp/src/ts/components/sender/sender.component.ts:
```diff
@@ -0,0 +1,17 @@
+
+  constructor() { }
+
+  ngOnInit(): void { }
```
Nope, it's the default from the semantics, I've removed it.

latin-panda commented on webapp/src/ts/modals/send-message/send-message.component.ts:
```diff
@@ -0,0 +1,221 @@
+    super();
+  }
+
+  ngOnInit(): void {}
```
Removed.

latin-panda commented on webapp/src/ts/modals/send-message/send-message.component.ts:
```diff
@@ -0,0 +1,221 @@
+    const type = contactTypes.find(type => type.id === typeId) || {};
+    let contact;
+
+    if (row.everyoneAt) {
```
Added back.

latin-panda replied:
Shouldn't this ToDos be a issue in Github instead? To follow up on them

dianabarsan replied:
yep, definitely! but if there are no issues and no todos, then it's lost forever :(

latin-panda commented on webapp/src/ts/modals/send-message/send-message.component.ts:
```diff
@@ -0,0 +1,221 @@
+      return this.formatPlace(row);
+    }
+
+    return row.doc.name || row.doc.phone;
```
Added back.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+        this.messagesActions.updateSelectedConversation({ contact: contactModel, messages: conversation });
+        this.globalActions.setTitle((contactModel && contactModel.doc && contactModel.doc.name) || id);
+        this.markConversationReadIfNeeded();
+        setTimeout(() => this.scrollToUnread());
```
@dianabarsan this code is running in the last "init" hook, so the next hook is one of the "change" hook, which will execute scrollToUnread on every little change in the template. 
If we try to do this in the init hooks, there's a change that the scroll wont happen on time if the request to retrieve conversations is slow for some reason.
I tried to chain it with promises, for example: after retrieving conversation then scroll, but this isn't giving enough time for the elements to be in the template and the scroll to find it.
I tried async await the selectContact and then scroll, but same bad result.
So, maybe if we don't want timeout here, we can put the scroll function in the "change" hook to let it try and try until it finds the element and then make it stop executing with a flag or something.

dianabarsan replied:
I think it's fine to execute in a change hook with a flag. 👍🏻 do you think it's worth the try?

latin-panda replied:
I tried to do that approach by using the 2 change hooks with a flag. It partially works, it's not scrolling all the way down, I think it's because it scrolls when it finds the wrapper container, but the content inside hasn't finished injecting all the items and grows a bit more.
I think we have 2 option: let it scroll by every template change (impacting performance), or use the timeout that I'm not too sure if it's going to be 100% reliable. I'd be nice to know when the content has finished rendering...

latin-panda replied:
I decided to leave the setTimeout to avoid adding on performance.

dianabarsan replied:
Didn't the `setTimeout` trigger an application-wide change detection ? 
https://stackoverflow.com/a/42972499/1865409

Wouldn't that be the worst hit on performance?

latin-panda replied:
I'm still unsure if the native setTimeout really triggers an application-wide change detection, I need to read more on how they intercept the native setTimeout.
I have made few changes to avoid the setTimeout. I added a fixed min-height to the async pipes, used the afterViewChange and added a flag to avoid running more than necessary the scrolling. However, if something async is added and this modifies the container height, then the scrolling might be affected :(

latin-panda commented on webapp/src/ts/modals/delete-doc-confirm/delete-doc-confirm.component.ts:
```diff
@@ -0,0 +1,55 @@
+  ngOnInit(): void { }
+
+  close() {
+    this.bsModalRef.hide();
```
Navigation added

latin-panda commented on webapp/src/ts/modals/send-message/send-message.component.ts:
```diff
@@ -0,0 +1,221 @@
+      return Promise.resolve();
+    }
+
+    const fieldLabel = this.translateService.instant('tasks.0.messages.0.message');
```
Refactored the code and removed promises.

latin-panda commented on webapp/src/ts/actions/messages.ts:
```diff
@@ -1,15 +1,23 @@
-  constructor(private store: Store) {}
+  constructor(
+    private store: Store,
+    private globalActions: GlobalActions
```
Oh yeah, 😩. Refactored.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.html:
```diff
@@ -0,0 +1,85 @@
+          <div class="controls">
+            <textarea name="message"
+              [(ngModel)]="send.message"
+              placeholder="{{ ('Reply to' | translate) + ' ' + (selectedConversation.contact?.doc?.name || selectedConversation.contact?.name || '') }}">
```
Fixed.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+      return;
+    }
+
+    const refreshSelected = this.selectedConversation && this.selectedConversation.id === id;
```
Fixed. So it removes the event and then adds it back again haha, but only if everything was successful.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+        const first = $('.item-content .body #message-content ul > li').filter(':first');
+        this.markConversationReadIfNeeded();
+
+        let scroll:any = false;
```
I think it should be find because this code executes when the $('.message-content-wrapper') exists, on db changes or on scroll event from the same element. It's not on the exact init time of the component .

dianabarsan replied:
It's happened a couple of times, when I was scrolling to the top to load more messages, that this code scrolled to the top sooner somehow and to load more messages I had to scroll lower and scroll up again. 

When that happened I also had this error in the console:
![image](https://user-images.githubusercontent.com/35681649/96552147-0c08ff80-12bc-11eb-94db-cba0c2fbffcc.png)

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+        this.textAreaFocused = true;
+      })
+      .on('blur', '#message-footer textarea', () => {
+        // Setting timeout to give change for clicking "add recipient" before hiding section.
```
It's hiding when the textarea looses focus, then it doesn't give time to click on the buttons. I think it makes more sense that we hide if the focus is lost from the container wrapping the textarea and the buttons instead of just textarea. What do you think about this idea?

dianabarsan replied:
You can give it a try, most important is to have the same behavior as before.

latin-panda replied:
I think it's fixed now. I had to refactored a bit.

latin-panda commented on webapp/src/ts/services/select2-search.service.ts:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  private initSelect2(selectEl, options, types) {
+    if (options.allowNew && types.length !== 1) {
```
Yes, that's good, I moved it to the parent just before calling this function

latin-panda commented:
> I'm not seeing the contact phone number when loading a conversation.

Fixed

latin-panda commented:
> 2. When selecting a really long conversation, scrolling to the top doesn't load more messages, I'm also not seeing "No more messages" when all messages are loaded.

Fixed.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+      ])
+      .then(([contactModel, conversation]) => {
+        if (!contactModel && conversation.length) {
+          const firstTaskWithContact = conversation[0].doc.tasks.find((task) => {
```
Fixed

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,322 @@
+            return message && message.contact && message.contact._id === id;
+          });
+
+          const firstMessageWithContact = firstTaskWithContact.messages.find(message => message.contact._id === id);
```
Fixed

latin-panda commented:
> I have a rather long LHS conversations list. When I select one and submit a message, I'm seeing the loader spinner when the LHS is updated, while on master, I don't see the spinner.

Fixed

latin-panda commented:
> 5. I'm seeing inconsistent results when searching in the select2search.

Fixed

latin-panda commented on webapp/tests/karma/ts/reducers/messages.spec.ts:
```diff
@@ -44,18 +42,105 @@ describe('Messages Reducer', () => {
+    const result = messagesReducer(state, action);
+
+    expect(result).to.deep.include(expectedState);
+  });
```
Done.

latin-panda commented on webapp/tests/karma/ts/services/send-message.spec.ts:
```diff
@@ -0,0 +1,331 @@
+      to: '+5552',
+      contact: { _id: recipient._id }
+    });
+    expect(markReadService.markAsRead.callCount).to.equal(1);
```
Done! Lovely tip tnx!

latin-panda commented on webapp/tests/karma/ts/services/send-message.spec.ts:
```diff
@@ -0,0 +1,331 @@
+    });
+    const task1Id = post.args[0][0].tasks[0].messages[0].uuid;
+    const task2Id = post.args[0][0].tasks[1].messages[0].uuid;
+    expect(task1Id).to.not.equal(task2Id);
```
Solved

latin-panda commented on webapp/tests/karma/ts/modules/messages/messages-content.component.spec.ts:
```diff
@@ -0,0 +1,196 @@
+      expect(updateSelectedConversationSpy.callCount).to.equal(0);
+      expect(removeMessageFromSelectedConversationSpy.callCount).to.equal(1);
+      expect(removeMessageFromSelectedConversationSpy.args[0]).to.deep.equal(['message2']);
+    });
```
Fixed!

latin-panda commented:
Hi @dianabarsan, I worked all the comments, could you please take a look again? 🙏 Thank you!

dianabarsan commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -0,0 +1,356 @@
+      this.hasToScroll = false;
+      this.allLoaded = false;
+      this.selectContact(this.urlParameters.id, this.urlParameters.type);
+      this.changeDetectorRef.detectChanges();
```
Is this still needed? I tried removing it locally and messages loaded correctly.

latin-panda replied:
Yes, I tried to remove it but I'm getting the exception about change detection :( 
I've put it back

dianabarsan commented on webapp/src/ts/modules/messages/messages-content.component.html:
```diff
@@ -0,0 +1,91 @@
+                  <p>
+                    {{message.doc?.sms_message?.message}}
+                    <span *ngIf="message.doc?.kujua_message">
+                      {{message.doc?.tasks[0]?.messages[0]?.message}}
```
It's possible for a message to have no tasks.

latin-panda replied:
Added extra check for tasks

dianabarsan commented on webapp/src/ts/modals/delete-doc-confirm/delete-doc-confirm.component.ts:
```diff
@@ -1,40 +1,72 @@
+    const contacts = 'contacts';
+    const reports = 'reports';
+
+    if (url.includes(contacts)) {
```
I think you might need to include `messages` here as well. 

If I load a conversation with one message and delete that message, I'm not redirected anywhere.

latin-panda replied:
The original code only redirect when it’s reports or contacts... if I add messages, then I need to listen to a selected conversation or pass over something from the global action that calls this modal and check if it's the last message before redirecting. I have added a [ToDo for improvement ](https://docs.google.com/spreadsheets/d/13BBw8cyV8CLMOzZAR_WDBqkaNrywNOk-i7b7pcypWJ8/edit#gid=303874733)

dianabarsan replied:
Cool!

dianabarsan commented on webapp/src/ts/modules/messages/messages-content.component.ts:
```diff
@@ -101,6 +105,14 @@ export class MessagesContentComponent implements OnInit, OnDestroy, AfterViewIni
+    const lastMessage = $('.message-content-wrapper .last-message');
+    // Scrolling only when last message is rendered.
+    if (this.hasToScroll && !this.loadingContent && lastMessage.length) {
+      this.scrollToUnread();
```
I'm not sure this is triggered correctly. I loaded a conversation with many read messages and was still scrolled all the way to the bottom :(

latin-panda replied:
It's hard to determine when the view has finished rendering so the component calculates the scroll correctly. Sadly, I had to add a timeout for now and added a [ToDo for improvement](https://docs.google.com/spreadsheets/d/13BBw8cyV8CLMOzZAR_WDBqkaNrywNOk-i7b7pcypWJ8/edit#gid=303874733)

dianabarsan reviewed (commented):
Really really nice! Thanks for all the work! 

I just found a few misses inline and one small request.

latin-panda commented on webapp/src/ts/modules/messages/messages-content.component.html:
```diff
@@ -0,0 +1,91 @@
+        <p class="loading-status" *ngIf="allLoaded" translate>No more messages</p>
+        <div>
+          <ul>
+            <li *ngFor="let message of selectedConversation.messages; trackBy: listTrackBy; let lastMessage = last" [ngClass]="{'outgoing': message.doc?.kujua_message, 'incoming': !message.doc?.kujua_message}">
```
Definitely. I was thinking about this while debugging the scrolling issue today. Added! :)

latin-panda commented:
@dianabarsan For some weird reason I couldn't reply to your comment: 
https://github.com/medic/cht-core/pull/6667#discussion_r508257528
We're having the same issue here with the scrolling where we need to determine when the view has finished rendering to calculate the scroll accurately. In this case the messages hasn't finished adding to the view. 
I added back the original timeout from master and added a ToDo 😩, definitely we need a clean better mechanisms for this.

dianabarsan reviewed (approved):
Awesome! Love it!
